micropython/lora/sx1276.py

```diff
@@ -0,0 +1,620 @@
+                _DETECT_OPTIMIZE_SF6 if sf == 6 else _DETECT_OPTIMIZE_OTHER,
+            )
+
+            modem_config2 |= sf << _MODEM_CONFIG2_SF_SHIFT
+            update_mask |= _MODEM_CONFIG2_SF_MAX << _MODEM_CONFIG2_SF_SHIFT
+            self._t_sym_ms = 1000 * (1 << sf) // self._bw_hz
+        except KeyError:
+            pass
+
+        try:
+            # TODO: make configurable. we only want this for TX mode, or RX in implicit header mode (only)
+            modem_config2 |= _MODEM_CONFIG2_RX_PAYLOAD_CRC_ON
+            update_mask |= _MODEM_CONFIG2_RX_PAYLOAD_CRC_ON
+        except KeyError:
+            pass
+
+        self._reg_update(_REG_MODEM_CONFIG2, update_mask, modem_config2)
+
+        # Update _REG_INVERT_IQ
+        invert_iq = 0x0
+        update_mask = 0x0
+        try:
+            invert_iq |= _INVERT_IQ_RX if lora_cfg["invert_iq_rx"] else 0
+            update_mask |= _INVERT_IQ_RX
+        except KeyError:
+            pass
+        try:
+            invert_iq |= _INVERT_IQ_TX if lora_cfg["invert_iq_tx"] else 0
+            update_mask |= _INVERT_IQ_TX
+        except KeyError:
+            pass
+        self._reg_update(_REG_INVERT_IQ, update_mask, invert_iq)
+
+        # TODO: Set Low Data Rate Optimization based on symbol rate (required if symbol duration > 16ms), semtech driver
+        # seems to set this for BW<=250kHz (TODO check)
+
+        try:
+            preamble_len = lora_cfg["preamble_len"]
+            preamble_len = bytearray([preamble_len >> 8, preamble_len])
+            self._reg_write(_REG_PREAMBLE_LEN_MSB, preamble_len)
+        except KeyError:
+            pass
+
+        # TODO: make AGC configurable?
+        self._reg_write(_REG_MODEM_CONFIG3, _MODEM_CONFIG3_AGC_ON)
+
+        # Come out of sleep mode
+        self._set_mode(_OPMODE_MODE_STDBY)
+
+    def _reg_write(self, reg, value):
+        if isinstance(value, int):
+            value = bytearray([value])
+        try:
+            self._nss(0)
+            self._spi.read(1, reg | _WRITE_REG_BIT)
+            self._spi.write(value)
+        finally:
+            self._nss(1)
+        if __DEBUG__:
+            print("W {:#x} ==> {}".format(reg, binascii.hexlify(value)))
+            self._reg_read(reg)  # log the readback as well
+
+    def _reg_update(self, reg, update_mask, new_value):
+        # Note: this function has a race condition, if the register value changes between
+        # read and write then old_value will be stale
+        old_value = self._reg_read(reg)
+        if update_mask:
+            value = ((old_value & ~update_mask) & 0xFF) | (new_value & update_mask)
+            self._reg_write(reg, value)
+        return old_value
+
+    def _reg_read(self, reg):
+        buf = bytearray(1)
+        self._reg_readinto(reg, buf)
+        return buf[0]
+
+    def _reg_readinto(self, reg, buf):
+        try:
+            self._nss(0)
+            self._spi.read(1, reg)
```

Comment:
To eliminate a memory allocation with this read, you can do:
```py
self._spi.readinto(self._buf1, reg)
```
where `self._buf1 = bytearray(1)` is in the constructor.